File: src/main/java/com/pedigree/editor/AlignAndDistributeController.java
package com.pedigree.editor;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;

/**
 * Implements alignment and distribution tools on selected nodes using a PositionAccessor.
 */
public class AlignAndDistributeController {

    public enum Alignment { TOP, MIDDLE, BOTTOM, LEFT, CENTER, RIGHT }
    public enum Distribution { HORIZONTAL, VERTICAL }

    public interface PositionAccessor {
        double getX(String id);
        double getY(String id);
        double getWidth(String id);
        double getHeight(String id);
        void setX(String id, double x);
        void setY(String id, double y);
    }

    private final SelectionModel selectionModel;
    private final PositionAccessor accessor;

    public AlignAndDistributeController(SelectionModel selectionModel, PositionAccessor accessor) {
        this.selectionModel = Objects.requireNonNull(selectionModel, "selectionModel");
        this.accessor = Objects.requireNonNull(accessor, "accessor");
    }

    public void align(Alignment mode) {
        List<String> ids = new ArrayList<>(selectionModel.getSelectedIds());
        if (ids.size() <= 1) return;

        double minX = ids.stream().mapToDouble(accessor::getX).min().orElse(0);
        double minY = ids.stream().mapToDouble(accessor::getY).min().orElse(0);
        double maxX = ids.stream().mapToDouble(id -> accessor.getX(id) + accessor.getWidth(id)).max().orElse(0);
        double maxY = ids.stream().mapToDouble(id -> accessor.getY(id) + accessor.getHeight(id)).max().orElse(0);
        double centerX = (minX + maxX) / 2.0;
        double centerY = (minY + maxY) / 2.0;

        for (String id : ids) {
            double x = accessor.getX(id);
            double y = accessor.getY(id);
            double w = accessor.getWidth(id);
            double h = accessor.getHeight(id);

            switch (mode) {
                case LEFT -> accessor.setX(id, minX);
                case RIGHT -> accessor.setX(id, maxX - w);
                case CENTER -> accessor.setX(id, centerX - w / 2.0);
                case TOP -> accessor.setY(id, minY);
                case BOTTOM -> accessor.setY(id, maxY - h);
                case MIDDLE -> accessor.setY(id, centerY - h / 2.0);
            }
        }
    }

    public void distribute(Distribution mode) {
        List<String> ids = new ArrayList<>(selectionModel.getSelectedIds());
        if (ids.size() <= 2) return;

        if (mode == Distribution.HORIZONTAL) {
            ids.sort(Comparator.comparingDouble(accessor::getX));
            double left = accessor.getX(ids.get(0));
            double right = accessor.getX(ids.get(ids.size() - 1)) + accessor.getWidth(ids.get(ids.size() - 1));
            double totalWidth = ids.stream().mapToDouble(accessor::getWidth).sum();
            double gap = (right - left - totalWidth) / (ids.size() - 1);

            double cursor = left;
            for (String id : ids) {
                accessor.setX(id, cursor);
                cursor += accessor.getWidth(id) + gap;
            }
        } else {
            ids.sort(Comparator.comparingDouble(accessor::getY));
            double top = accessor.getY(ids.get(0));
            double bottom = accessor.getY(ids.get(ids.size() - 1)) + accessor.getHeight(ids.get(ids.size() - 1));
            double totalHeight = ids.stream().mapToDouble(accessor::getHeight).sum();
            double gap = (bottom - top - totalHeight) / (ids.size() - 1);

            double cursor = top;
            for (String id : ids) {
                accessor.setY(id, cursor);
                cursor += accessor.getHeight(id) + gap;
            }
        }
    }
}

File: src/main/kotlin/com/family/tree/editor/AlignAndDistributeController.kt
package com.family.tree.editor

import java.util.*
import kotlin.math.max
import kotlin.math.min

/**
 * Implements alignment and distribution tools on selected nodes using a PositionAccessor.
 */
class AlignAndDistributeController(
    private val selectionModel: SelectionModel,
    private val accessor: PositionAccessor
) {
    init {
        Objects.requireNonNull(selectionModel, "selectionModel")
        Objects.requireNonNull(accessor, "accessor")
    }

    enum class Alignment { TOP, MIDDLE, BOTTOM, LEFT, CENTER, RIGHT }
    enum class Distribution { HORIZONTAL, VERTICAL }

    interface PositionAccessor {
        fun getX(id: String): Double
        fun getY(id: String): Double
        fun getWidth(id: String): Double
        fun getHeight(id: String): Double
        fun setX(id: String, x: Double)
        fun setY(id: String, y: Double)
    }

    fun align(mode: Alignment) {
        val ids = ArrayList(selectionModel.getSelectedIds())
        if (ids.size <= 1) return

        val minX = ids.minOfOrNull { accessor.getX(it) } ?: 0.0
        val minY = ids.minOfOrNull { accessor.getY(it) } ?: 0.0
        val maxX = ids.maxOfOrNull { accessor.getX(it) + accessor.getWidth(it) } ?: 0.0
        val maxY = ids.maxOfOrNull { accessor.getY(it) + accessor.getHeight(it) } ?: 0.0
        val centerX = (minX + maxX) / 2.0
        val centerY = (minY + maxY) / 2.0

        for (id in ids) {
            val x = accessor.getX(id)
            val y = accessor.getY(id)
            val w = accessor.getWidth(id)
            val h = accessor.getHeight(id)

            when (mode) {
                Alignment.LEFT -> accessor.setX(id, minX)
                Alignment.RIGHT -> accessor.setX(id, maxX - w)
                Alignment.CENTER -> accessor.setX(id, centerX - w / 2.0)
                Alignment.TOP -> accessor.setY(id, minY)
                Alignment.BOTTOM -> accessor.setY(id, maxY - h)
                Alignment.MIDDLE -> accessor.setY(id, centerY - h / 2.0)
            }
        }
    }

    fun distribute(mode: Distribution) {
        val ids = ArrayList(selectionModel.getSelectedIds())
        if (ids.size <= 2) return

        if (mode == Distribution.HORIZONTAL) {
            ids.sortBy { accessor.getX(it) }
            val left = accessor.getX(ids.first())
            val last = ids.last()
            val right = accessor.getX(last) + accessor.getWidth(last)
            val totalWidth = ids.sumOf { accessor.getWidth(it) }
            val gap = (right - left - totalWidth) / (ids.size - 1)

            var cursor = left
            for (id in ids) {
                accessor.setX(id, cursor)
                cursor += accessor.getWidth(id) + gap
            }
        } else {
            ids.sortBy { accessor.getY(it) }
            val top = accessor.getY(ids.first())
            val last = ids.last()
            val bottom = accessor.getY(last) + accessor.getHeight(last)
            val totalHeight = ids.sumOf { accessor.getHeight(it) }
            val gap = (bottom - top - totalHeight) / (ids.size - 1)

            var cursor = top
            for (id in ids) {
                accessor.setY(id, cursor)
                cursor += accessor.getHeight(id) + gap
            }
        }
    }
}
